model/websocket_client.go

```diff
@@ -259,21 +263,27 @@ func (wsc *WebSocketClient) pingHandler(appData string) error {
 	if atomic.LoadInt32(&wsc.closed) == 1 {
 		return nil
 	}
-	if !wsc.pingTimeoutTimer.Stop() {
-		<-wsc.pingTimeoutTimer.C
-	}
-
-	wsc.pingTimeoutTimer.Reset(time.Second * (60 + PING_TIMEOUT_BUFFER_SECONDS))
+	wsc.resetTimerChan <- struct{}{}
 	wsc.writeChan <- writeMessage{
 		msgType: msgTypePong,
 	}
 	return nil
 }
 
 func (wsc *WebSocketClient) pingWatchdog() {
-	select {
-	case <-wsc.pingTimeoutTimer.C:
-		wsc.PingTimeoutChannel <- true
-	case <-wsc.quitPingWatchdog:
+	for {
+		select {
+		case <-wsc.resetTimerChan:
+			if !wsc.pingTimeoutTimer.Stop() {
+				<-wsc.pingTimeoutTimer.C
+			}
+			wsc.pingTimeoutTimer.Reset(time.Second * (60 + PING_TIMEOUT_BUFFER_SECONDS))
+
+		case <-wsc.pingTimeoutTimer.C:
+			wsc.PingTimeoutChannel <- true
```

Comment:
Just noting that this could potentially block indefinitely if `PingTimeoutChannel` is not correctly handled by the user of the API.
Whether we should adopt a "better safe than sorry" approach and wrap this in a `select` or just make the expected usage explicit through documentation, I will leave it to you.

Reply:
This is complicated as is. I don't believe somebody is using the websocket client as heavily as us. And the API itself is error-prone which should be changed in v6. Let's just document this for now.